sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -322,6 +323,62 @@ class Analyzer(
   }
 
   /**
+    * Replace the index with the related attribute for order by and group by.
+    */
+  object ResolveIndexReferences extends Rule[LogicalPlan] {
+
+    def indexToColumn(
+        sortOrder: SortOrder,
+        child: LogicalPlan,
+        index: Int,
+        direction: SortDirection): SortOrder = {
+      val orderNodes = child.output
+      if (index > 0 && index <= orderNodes.size) {
+        SortOrder(orderNodes(index - 1), direction)
+      } else {
+        throw new UnresolvedException(sortOrder,
+          s"""Order by position: $index does not exist \n
+              |The Select List is indexed from 1 to ${orderNodes.size}""".stripMargin)
+      }
+    }
+
+    def indexToColumn(
+        agg: Aggregate,
+        child: LogicalPlan,
+        index: Int): Expression = {
+      val attributes = child.output
+      if (index > 0 && index <= attributes.size) {
+        attributes(index - 1)
+      } else {
+        throw new UnresolvedException(agg,
+          s"""Aggregate by position: $index does not exist \n
+              |The Select List is indexed from 1 to ${attributes.size}""".stripMargin)
+      }
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case s @ Sort(orders, global, child) if child.resolved =>
```

Comment:
That's exactly what I trying to achieve. The thing is `expression` is hided within Seq[SortOrder], somehow we need to iterate `order` to do the checking and the conversion if match. Cann't see the chance to compact them into one line condition, and there's no performance difference in this way. 
``` scala
case class Sort(
    order: Seq[SortOrder],
    global: Boolean,
    child: LogicalPlan)
case class SortOrder(child: Expression, direction: SortDirection)
```